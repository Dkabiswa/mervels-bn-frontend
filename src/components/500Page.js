import React from 'react';

function ServerErrorPage() {
    return (
        <div className="c4v-contianer">
<div className="c4v">
        <img alt="barefootNomad Logo" className="barefoot-logo" src="https://res.cloudinary.com/bahati/image/upload/v1573114920/marvel_logo_fngq4h.png"/>
        <div className="call4verif-container">
            <h3 className="c4v-title">SERVER ERROR</h3>
            <h1 className="error-num">500</h1>
            <p>Problem Occured Try Again OR Check Your Internet Connection</p>
        </div>
        </div>
        </div>
);
}

export default ServerErrorPage;