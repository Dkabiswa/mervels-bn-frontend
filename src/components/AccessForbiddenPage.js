
/* eslint-disable no-unused-vars */
import React from 'react';
import "../styles/403.scss";
import "../styles/App.scss";

function AccessForbiddenPage(props) {
    return (
        <>
            <div id="AccessForbidden">
                <div>403</div>
                <div className="txt">
                    Forbidden<span className="blink">_</span>
                </div>
                <div className="txt m-top-2">
                    <a href="/home">Go home</a>
                </div>
            </div>
        </>
    );
}

export default AccessForbiddenPage;

