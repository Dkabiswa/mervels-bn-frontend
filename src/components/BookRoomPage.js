/* eslint-disable no-unused-vars */
/* eslint-disable no-debugger */
/* eslint-disable no-unneeded-ternary */
/* eslint-disable no-nested-ternary */
/* eslint-disable react/no-unused-state */
/* eslint-disable no-shadow */
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/require-default-props */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable import/no-unresolved */
import React, {Component} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {toast} from 'react-toastify';
import { Helmet } from "react-helmet";
import bookRoom from '../redux/actions/bookingAction';
import BookCard from './shared/BookCard';
import {getAccommodations, getAccommodation } from '../redux/actions/accommodationsAction';
import AccommodationCard from './shared/accommodationCard';
import getRooms from '../redux/actions/roomsActions';
import {validateBooking} from '../helpers/validator';
import Meta from './shared/meta';

class BookRoom extends Component {

    constructor(props) { 
        super(props);
        this.state = {
            activeLocation: undefined,
            error: {
                details: '',
                class: 'hidden'
            }
        };
    }

    componentWillMount() {
        const {request} = this.props;
        const { accommodations } = request.data[0];
        const oneReq = request.data[0];
        const requests = [];
        accommodations.forEach((acc, index) => {
            const req1 = {};
            const {travelDate, returnDate, accommodations} = oneReq;
            req1.travelDate = travelDate[index];
            req1.returnDate = returnDate;
            req1.accommodation = accommodations[index].name;
            req1.location = accommodations[index].Location;
            req1.accommodationId = accommodations[index].id;
            requests.push(req1);
        });
        this.setState({trips: requests});
        const { getAccommodations, user, getRooms } = this.props;
        getAccommodations({role: user });
        getRooms();
    }

    async componentWillReceiveProps(nextProps) {
        const {bookedroom, history} = nextProps;
        if(bookedroom.data || bookedroom.error){
            const {data, error} = bookedroom;
            if(data) {
                toast.success(data.message);
                history.push('/requests');
            } else {
                const {status, message} = error;
                if(status===422) {
                    toast.error("Check your inputs");
                } else if(status === 401) {
                    toast.error("The current session is expired. Login");
                    localStorage.removeItem('bareFootToken');
                    history.push('/login');
                } else if(status === 403) {
                    toast.error(message);
                    history.push('/dashboard');
                } else if(status === 404) {
                    toast.error(message);
                    history.push('/requests');
                } else if(status === 409){
                    toast.error(message);
                    history.push('/requests');
                } else {
                    toast.error(message);
                }
            }
        }
    }

    handleImageClick = async (id, name) => {
        const {activeLocation, trips} = this.state;
        const value = name;
        if(activeLocation === undefined) {
            const activeLoc = trips[0].location.id;
            this.setState({activeLocation: activeLoc});
        }

        // find whic array contsins the location
        
        let parent;
        trips.forEach((trip, index) => {
            if(trip.location.id === activeLocation) {
                parent = index;
            }
        });

        trips[parent] = {...trips[parent], accommodation: value};
        this.setState({trips});
            const {Allaccommodations, allRooms} = this.props;
            const accommodation = Allaccommodations.accommodations.find((acc) => acc.name === value);
            const accommodationId = accommodation.id;
            const availableRooms = allRooms.data.filter((room) => room.accommodationId === accommodationId);
            trips[parent] = {...trips[parent], accommodationId, availableRooms};
            this.setState({trips});  
    }

    handleChange = async ({target}) => {
        const parent =  target.attributes.unique.value;
        const {trips} = this.state;
        const {value} = target;
        if(target.name === 'accommodation' && value !== 'Select Accommodation') {
            trips[parent] = {...trips[parent], [target.name]: value};
            this.setState({trips});
            this.setState({selectedAcommodation: value});
            const {Allaccommodations, allRooms} = this.props;
            const accommodation = Allaccommodations.accommodations.find((acc) => acc.name === value);
            const {id} = accommodation;
            const availableRooms = allRooms.data.filter((room) => room.accommodationId === id);
            trips[parent] = {...trips[parent], accommodationId: id, availableRooms};
            this.setState({trips});
        } else if(target.name==='room') {
            const {activeLocation, trips, possibleAccommodations} = this.state; // active Location is shitting me.
            const selectedAcommodation = trips[parent].accommodation;
            const {allRooms} = this.props;
            const activeAccommodation = possibleAccommodations.find((acc)=> acc.locationId === activeLocation && acc.name === selectedAcommodation); // should use possibleAccommodations
            const {id} = activeAccommodation; // this might be the mistake
            if(value !== 'Select Room') {
            const activeRoom = allRooms.data.find((room) => {
                return room.accommodationId === id && room.name === value;
            });          
            const roomId = activeRoom.id;
            trips[parent] = {...trips[parent], [target.name]: roomId};
            this.setState({trips});
            }
        } else {
            trips[parent] = {...trips[parent], [target.name]: value};
            this.setState({trips});
        }
        
    }


    handleFocus = (locationId, possibleAccommodations) => {
        const {activeLocation} = this.state;
        if(locationId!==activeLocation) {
            this.setState({activeLocation: locationId, possibleAccommodations});
        }
    }

    handleCancel = () => {
       const {toggleBooking} = this.props;
       toggleBooking();
    }

    handleSubmit = async () => {

        const {trips} = this.state;
        const {bookRoom, request} = this.props;
        const requestId = request.data[0].id;
        const bookDetails =[];
        const toValidate =[];
        trips.forEach((trip) => {
            const {checkIn, checkOut, accommodation, room, travelDate, returnDate} = trip;
            bookDetails.push({checkIn, checkOut, accommodation, room});
            toValidate.push({checkIn, checkOut, accommodation, room, travelDate, returnDate}); 
        });
        const error = validateBooking(toValidate);
        if(!error) {
            bookRoom(bookDetails , requestId);
        } else {
            this.setState({error: {details: error, class: 'visible'}});
        }

    }

    render() {
        const {classes, Allaccommodations, allRooms } = this.props;
        const nonBookedRooms = allRooms.data.filter((room) => room.status === 'Available');
        const {trips, possibleAccommodations, availableRooms, selectedAcommodation, error} = this.state;
        const acommodations = Allaccommodations.accommodations.filter((acc) => acc.status === 'Available');
        const fLocId = trips[0].location.id;
        const firstAccommodations= acommodations.filter((acc) => acc.locationId === fLocId );
        const selectedRooms = [];
        trips.forEach((trip)=>{
            selectedRooms.push(trip.room);
        });
        const filteredAcoomodations = possibleAccommodations || firstAccommodations;

        return ( <div className = {`${classes}`} >
                <Meta  title="Booking-Rooms"/>
                <div className="bg-img" />
                <div className="black-container black-short" />
                <p className='accommodation-title p-top-5 m-bottom-3'>Booking</p>
                <div className = "bookcards-container">
                    {trips.map((trip, index) => (<BookCard trips={trips} handleGetRooms={this.handleGetRooms} selectedRooms={selectedRooms} availableRooms={availableRooms} displayRooms={this.displayRooms} rooms ={nonBookedRooms} handleFocus={this.handleFocus} key={trip.location.id} locationId={trip.location.id} handleChange={this.handleChange} parent={index} accommodations={acommodations} accommodation={selectedAcommodation || trip.accommodation} returnDate={trip.returnDate} travelDate={trip.travelDate} destination={`${trip.location.city} ${trip.location.country}`}/>))}
                </div>
                <div className={`${error.class} book-error error-multi-city`}>
                    {error.details}
                </div>
                <div className = "book-buttons-container">
                    <button id="cancel-booking" type = "button"className = "btn btn-danger" onClick={this.handleCancel}> Cancel </button>
                    <button id="submit-booking" type = "submit" className = "btn btn-secondary" onClick={this.handleSubmit}> Submit </button> 
                </div>
                <div className = "accommodations-container">
                    {filteredAcoomodations.map((acc) => (<AccommodationCard handleImageClick={this.handleImageClick} name = {acc.name} url={acc.imageUrl} id={acc.id}/>))}
                </div>
            </div>
        );
    }
};

BookRoom.propTypes = {
    request: PropTypes.object.isRequired,
    classes: PropTypes.string.isRequired,
    getAccommodations: PropTypes.func.isRequired,
    Allaccommodations: PropTypes.object.isRequired,
    user: PropTypes.string.isRequired,
    getRooms: PropTypes.func.isRequired,
    allRooms: PropTypes.object.isRequired,
    toggleBooking: PropTypes.object.isRequired,
    bookRoom: PropTypes.func.isRequired,
    bookedroom: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
};

const mapStateToProps = ({ request, accommodations, accommodation, profile, allRooms, bookedroom }) => {
    return {
        request,
        Allaccommodations: accommodations,
        accommodation,
        user: profile.data.role,
        allRooms,
        bookedroom
    };
};

const mapDispatchToProps = {
    bookRoom,
    getAccommodations,
    getAccommodation,
    getRooms,
};

export {BookRoom};

export default connect(mapStateToProps, mapDispatchToProps)(BookRoom);