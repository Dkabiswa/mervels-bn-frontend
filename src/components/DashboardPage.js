/* eslint-disable react/prop-types */
/* eslint-disable react/self-closing-comp */
import React from "react";
// import Ratings from "./shared/Rating";
import Button from "./shared/Button";
import Select from "./shared/Select";
// import imagePlaceholder from "../assets/hotel_one.jpg";
import Input from "./shared/input";
import SingleAccommodation from './accommodation';
import { enhanceRooms } from '../helpers/RoomClasses';

class DashboardPage extends React.Component {
  render() {
    const { destination, trips, computePeriod, updateInput } = this.props;
    let single;
    let rooms;
    if(destination.count > 0) {
      const accommodation = destination.destinations[0];
      const location = `${accommodation.Location.city} ${accommodation.Location.country}`;
      single = <SingleAccommodation 
        name={accommodation.name}
        likes={accommodation.Likes.length}
        rooms={accommodation.Rooms.length}
        rating = {accommodation.Ratings.averageRating}
        location={location}
        id={accommodation.id}
        key={accommodation.id}
        imageUrl={accommodation.imageUrl}
        description={accommodation.description}
      />;
      const displayRooms = accommodation.Rooms.slice(0, 2);
      rooms = enhanceRooms(displayRooms);
    }
    return (
      <>
      <div className="container">
        <div className="grid grid-sm">
          <div className="col-6 offset-2 col-sm-12">
            <div className="dash-dest">
              <span className="card-title">MOST TRAVELLED DESTINATION</span>
              <div className="card-body">
                {single}
              </div>
            </div>
          </div>
          <div className="col-5 col-sm-12 dash-dest">
          <div className="dash-rooms">
            <div className="rooms">
                {rooms}
            </div>
        </div>
          </div>
          <div className="col-12" />
        </div>
        <div className="grid grid-sm dash-bottom">
        <div className="col-8 offset-4 col-sm-12 center dash-stat">
              <div className="card">
              <span className="card-title">Find out your past trips statistics</span>
              <div className="filter">
                <div className="filter-items">
                  <Select
                    name="parameter"
                    onChange={updateInput}
                    selected="years"
                    ids={["days", "weeks", "months", "years"]}
                    options={["days", "weeks", "months", "years"]}
                    error=""
                  />
                </div>
                <div className="filter-items">
                  <Input
                    name="value"
                    onChange={updateInput}
                    inputType="number"
                    placeholder="Enter Value"
                  />
                </div>
                <div className="filter-items">
                  <Button
                    ButtonId="compute"
                    onClick={computePeriod}
                    buttonType="button"
                    text="SUBMIT"
                    classes="btn btn-primary"
                  />
                </div>
              </div>
            </div>
          </div>
          <div className="col-4" />
          <div className="col-12 col-sm-11 offset-1">
          <span className="card-title">PREVIOUS TRIPS: {trips.total}</span>
            <div className="requests scroll_container">
              {trips.trips.length > 0 ? trips.trips.map(trip => (
                <div className="one_card reqcard" key={trip.id}>
                  <div className="card-info">
                    <div className="card-info-item">
                      <span className="label">Request date </span>
                      <span className="label-item">
                        {new Date(trip.createdAt).toLocaleDateString()}
                      </span>
                    </div>
                    <div className="card-info-item">
                      <span className="label">From </span>
                      <span className="label-item">{trip.from}</span>
                    </div>
                    <div className="card-info-item">
                      <span className="label">TravelDate </span>
                      <span className="label-item">{trip.travelDate}</span>
                    </div>
                    <div className="card-info-item">
                      <span className="label">ReturnDate </span>
                      <span className="label-item">{trip.returnDate}</span>
                    </div>
                    <div className="card-info-item">
                      <span className="label">Status </span>
                      <span className="label-item">
                        {trip.status}
                        <span
                          id="circle"
                          className={
                            trip.status === "Approved" ? "approved" : "rejected"
                          }
                        ></span>
                      </span>
                    </div>
                  </div>
                </div>
              )): <span className="text-center text-danger card-title">NO TRIPS</span>}
            </div>
          </div>
        </div>
      </div>
      </>
    );
  }
}

export default DashboardPage;
