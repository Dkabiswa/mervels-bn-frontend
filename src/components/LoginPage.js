/* eslint-disable react/no-did-update-set-state */
/* eslint-disable import/no-unresolved */
/* eslint-disable no-shadow */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable react/forbid-prop-types */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Button from './shared/Button';
import SocialAuth from "./shared/socialAuth";
import Input from "./shared/input";
import { localAuth, socialAuth } from "../redux/actions/login";
import logo from "../logo/logo@2x.png";

export class LoginPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
      submitting: false,
      error: ''
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (props.user.error !== state.error) {
      return {
        submitting: false,
        error: props.user.error
      };
    }
    return null;
  }

  componentDidMount() {
    this.setState(prevState => ({ ...prevState }));
    this.checkLoggedIn();
    const { location, socialAuth } = this.props;
    if (location !== undefined) {
      const base64encoded = location.search.split("&")[0].split("?code=")[1];
      if (base64encoded) {
        const decoded = JSON.parse(atob(base64encoded));
        socialAuth(decoded);
      }
    }
  }

  componentWillReceiveProps(nextProps) {
    const { user } = nextProps;
    if(user.error === 'Invalid email or password entered') {
      this.setState({ email: '', password: '' });
    }
  }

  componentDidUpdate() {
    const { user } = this.props;
    if (user.isLoggedIn === true) {
      const token = localStorage.getItem("bareFootToken");
      if (token) {
        window.location.href = '/';
      }
    }
  }

  checkLoggedIn = () => {
    const token = localStorage.getItem("bareFootToken");
    if (token !== null) {
      window.location.href = '/';
    }
  };

  handleInput = async e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  handleSubmit = e => {
    const { localAuth } = this.props;
    const { email, password } = this.state;
    this.setState({
      submitting: true,
      error: ''
    });
    e.preventDefault();
    const user = {
      userEmail: email,
      userPassword: password
    };
    localAuth(user);
  };

  render() {
    const { password, email, submitting } = this.state;
    return (
      <div className="login-container">
        <div className="local bg-white">
          <img src={logo} alt="logo" />
          <form className="loginForm" onSubmit={this.handleSubmit}>
            <Input
              name="email"
              inputType="email"
              placeholder="Email"
              onChange={this.handleInput}
              value={email}
              required
            />
            <Input
              name="password"
              inputType="password"
              placeholder="Password"
              onChange={this.handleInput}
              value={password}
              required
            />
            <div className="forgot">
              <a href="/forgotPassword" className="other-link">Forgot your password?</a>
            </div>
            <Button ButtonId='login' classes='btnn btnn-primary log' text='LOGIN' buttonType='submit' submitting={submitting} onClick={this.handleSubmit} />
          </form>
          <div className="social">
            <SocialAuth />
            <div className="foot-message">
          Dont have a Barefoot Nomad account? <a href="/signUp"><span className="other-link">Sign up now!</span></a>
          </div>
          </div>
        </div>
      </div>
    );
  }
}

LoginPage.propTypes = {
  localAuth: PropTypes.func.isRequired,
  socialAuth: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  // eslint-disable-next-line react/require-default-props
  location: PropTypes.object
};

const mapStateToProps = ({ loginReducer }) => ({
  user: loginReducer
});

export default connect(
  mapStateToProps,
  { localAuth, socialAuth }
)(LoginPage);
