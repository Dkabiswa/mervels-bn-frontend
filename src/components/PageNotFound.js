import React from 'react';

function PageNotFound() {
    return (<div className="c4v-contianer">
<div className="c4v">
        <img alt="barefootNomad Logo" className="barefoot-logo" src="https://res.cloudinary.com/bahati/image/upload/v1573114920/marvel_logo_fngq4h.png"/>
        <div className="call4verif-container">
            <h3 className="c4v-title">Page Not Found</h3>
            <h1 className="error-num">404</h1>
        </div>
        </div>
        </div>
);
}

export default PageNotFound;