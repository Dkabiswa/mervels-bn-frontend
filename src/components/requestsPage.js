/* eslint-disable react/prop-types */
/* eslint-disable no-lonely-if */
/* eslint-disable default-case */
/* eslint-disable no-shadow */
/* eslint-disable react/forbid-prop-types */
import React from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import TableComponent from './shared/tableComponent';
import { fetchRequests, getPending, getPast, searchRequests }  from '../redux/actions/requestsAction';
import 'react-toastify/dist/ReactToastify.css';
import RequestPanel from './headerPanel';
import { searchMessage } from '../helpers/search';



class RequestView extends React.Component{
    constructor(props) {
        super(props);
        this.state = {
            currentPage: 1,
            requestsPerPage: 2,
            parameter: null,
            query: null
        };
    }

    componentDidMount(){
        const { fetchRequests: getRequests } = this.props;
        getRequests();
    }

    componentDidUpdate(prevProps) {
        // eslint-disable-next-line react/prop-types
        const { requests, history } = this.props;
        if (requests !== prevProps.requests) {
            if(requests.error){
                if(requests.error.status === 401){
                        // eslint-disable-next-line react/prop-types
                        history.push('/login');
                }
            }
        }
    }

    handleChange = (e) => {
        this.setState({
            [e.target.name]: e.target.value
        });
    }

    handleRequests = (value) => {
      const { fetchRequests: getRequests, getPending, getPast } = this.props;
      switch (value) {
        case 'ALL':
          getRequests(); 
          break;  
        case 'PENDING':
          getPending(); 
          break; 
        case 'PAST':
          getPast(); 
          break; 
      }
    }

    setCurrentPage = (pageNumber) =>{
        this.setState({
            currentPage: pageNumber,
        });
    }

    paginate = (pageNumber) => this.setCurrentPage(pageNumber)

    handleSearch = () => {
        const { parameter, query } = this.state;
        const { searchRequests } = this.props;
        if(parameter && query ) {
            searchRequests(parameter, query);
        }
    }

    render(){
        const { parameter, currentPage } = this.state;
        let { requestsPerPage } = this.state;
        if(!requestsPerPage || requestsPerPage <= 0) {
            requestsPerPage = 1;
        }
        const { requests, match } = this.props;
        const { message, type } = searchMessage(parameter);
        let items;
        let display;

        const indexOfLastRequest = currentPage * requestsPerPage;
        const indexOfFirstRequest = indexOfLastRequest - requestsPerPage;
        display= <div>
                    <h2 className="text-center">Data Loading ...</h2>
                </div>;
        let obj;
        if (Object.keys(requests.filtered).length !== 0){
            obj = requests.filtered;
        }else {
            obj = requests.requests;
        }
        
        if(requests.error !== null){
            display = <div>
               <h3 className="text-center text-danger">{requests.error.status === 401 ? "Invalid Session" :requests.error.message}</h3>
         </div>;
        }else {
                if(Object.keys(obj).length !== 0 && obj.constructor === Object) {
                items =obj.data;
                const currentRequests = items.slice(indexOfFirstRequest, indexOfLastRequest);
                const table = <TableComponent
                items={currentRequests}
                route={match.path}
                currentPage={currentPage}
                requestsPerPage={requestsPerPage}
                totalRequests={items.length}
                paginate={this.paginate}
                />;

                display = items && items.length ? table :<p className="text-center backdrop">you currently have no requests</p>;
            }
        }

        return(
            <>
                <RequestPanel
                    title={requests.title} 
                    onClick={this.handleRequests} 
                    onChange={this.handleChange} 
                    holder={message} type={type} 
                    search={this.handleSearch}
                />
                <div className="grid sm p-bottom-3">
                    <div className="col-10 offset-3">
                    <div className="page-header text-center"  />
                   {display}
                    </div>
                </div>
            </>
        );
    }
}
RequestView.propTypes ={
    fetchRequests: propTypes.func.isRequired,
    searchRequests: propTypes.func.isRequired,
    requests: propTypes.object.isRequired,
    getPending: propTypes.func.isRequired,
    getPast: propTypes.func.isRequired,
};
export const mapStateToProps = (state) => {
    return {
        requests: state.requests,
        error: state.error
    };
};

const mapDispatchToProps = {
    fetchRequests,
    getPending,
    searchRequests,
    getPast
};

export { RequestView as RequestViewTest };
export default connect(mapStateToProps, mapDispatchToProps)(RequestView);