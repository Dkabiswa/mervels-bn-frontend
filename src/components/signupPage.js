/* eslint-disable react/require-default-props */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable import/no-unresolved */
import React, { Component } from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {toast} from 'react-toastify';
import signup from '../redux/actions/signupActions';
import Input from './shared/input';
import validator from '../helpers/validator';
import SocialAuth from './shared/socialAuth';

class SignUpPage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            errors: {
                firstName: undefined,
                lastName: undefined,
                userEmail: undefined,
                userPassword: undefined,
                confirm: undefined,
                match: undefined
            }
        };
    }

    async componentWillReceiveProps(nextProps) {
        if (nextProps.user.data !==null) {
            nextProps.history.push("/call4verify");
        } else {
            const {error} =  nextProps.user;
            toast.error(error.message);

        const button = await document.querySelector('button');
        button.innerHTML = "Sign Up";
        }
    }

    handleChange = async ({target}) => {
        this.setState((prev) => ({...prev, [target.name]: target.value}));
        let { error } = await validator(target.name, target.value);
        if(target.name==='confirm'){
            if(this.state.userPassword !== this.state.confirm){
                error = 'Passwords do not match';
            }
        }
        this.setState((prev) => ({...prev, errors: { ...prev.errors, [target.name]: error } }));
    };

    handleSubmit = async (event) => {
        event.preventDefault();
        const hasErrors = Object.values(this.state.errors).some(val=> val!==undefined);
        if (!hasErrors) {
            const signUp = this.props.signup;
            const userCredentials = this.state;
            signUp(userCredentials);
            const button = await document.querySelector('.signup-btn');
            button.innerHTML = "Wait ...";
        }
    }

    render() {
        const {firstName, lastName, userEmail, userPassword, confirm, errors} = this.state;
        return (
            <div className="sm-id">
                <div className="signup-page p-top-7 ">
                <div className="grid bg-white p-bottom-2">
                    <div className="col-2" />
                    <div className="col-4">
                        <img alt="barefootNomad Logo" className="col-12 barefoot-logo" src="https://res.cloudinary.com/bahati/image/upload/v1573114920/marvel_logo_fngq4h.png"/>
                        <form className="signup-form " onSubmit = {this.handleSubmit}>
                            <Input placeholder="First Name" error={errors.firstName} required={{required: 'required'}} onChange={this.handleChange} name="firstName" classes="input" inputType="text" value={firstName}/>
                            <Input placeholder="Last Name" error={errors.lastName} required={{required: 'required'}} onChange={this.handleChange} name="lastName" classes="input" inputType="text" value={lastName}/>
                            <Input placeholder="Email" error={errors.userEmail} required={{required: 'required'}} onChange={this.handleChange} name= "userEmail" classes="input" inputType="email" value={userEmail}/>
                            <Input placeholder="Password" error={errors.userPassword} required={{required: 'required'}} onChange={this.handleChange} name="userPassword" classes="input" inputType="password" value={userPassword}/>
                            <Input placeholder="Confirm Password" error={errors.confirm} required={{required: 'required'}} onChange={this.handleChange} name="confirm" classes="input" inputType="password" value={confirm}/>
                            <button type="submit" className="btn btn-primary signup-btn">Sign Up</button>
                        </form>
                    </div>
                    <div className='col-1 p-left-5 p-top-3' style={{textAlign: 'center'}}>
                        <div className="hr-vertical"/>
                    </div>
                    <div className="col-3 sm-1">
                        <SocialAuth/>
                        <div className="foot-message">Already have a Barefoot Nomad account? <a href="/login"><span className="other-link">Login</span></a></div>
                    </div>
                    <div className="col-2" />
                </div>
            </div>
            </div>
             );
    }
};

SignUpPage.propTypes = {
    signup: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    history: PropTypes.object,
    push: PropTypes.func
};

const mapStateToProps = ({user}) => {
    return {
        user
    };
};

const mapDispatchToProps = {
    signup
};

export { SignUpPage };
export default connect(mapStateToProps, mapDispatchToProps)(SignUpPage);