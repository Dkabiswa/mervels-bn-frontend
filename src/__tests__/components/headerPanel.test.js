/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { HeaderPanel } from '../../components/headerPanel';

const props = {
    title: "", 
    onClick: jest.fn(),
};
let wrapper;
describe('Header Panel Component', () => {
    test('should render', () => {
      wrapper = shallow(<HeaderPanel {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
    test('should call the onClick function "all"', () => {
        const mockOnClickFn = jest.fn();
        wrapper = mount(<HeaderPanel onClick={mockOnClickFn} title="ALL REQUESTS" />);
        wrapper.find('#all').simulate(
          'click'
        );
        expect(mockOnClickFn.mock.calls.length).toBe(1);
    });
    test('should call the onClick function "pending"', () => {
        const mockOnClickFn = jest.fn();
        wrapper = mount(<HeaderPanel onClick={mockOnClickFn} title="PENDING REQUESTS" />);
        wrapper.find('#pending').simulate(
          'click'
        );
        expect(mockOnClickFn.mock.calls.length).toBe(1);
    });
    test('should call the onClick function "past"', () => {
        const mockOnClickFn = jest.fn();
        wrapper = mount(<HeaderPanel onClick={mockOnClickFn} title="PAST REQUESTS" />);
        wrapper.find('#past').simulate(
          'click'
        );
        expect(mockOnClickFn.mock.calls.length).toBe(1);
    });
});