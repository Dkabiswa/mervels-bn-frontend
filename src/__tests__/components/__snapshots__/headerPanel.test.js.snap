// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Panel Component should render 1`] = `ShallowWrapper {}`;
